### ISSUE TITLE
SYN received in invalid state

### ISSUE THREAD

**Issue by Meatballs1** [iagox86/dnscat2#19]
Hi, I get the following repeated errors from a session:

```
[WARNING] INCOMING: [0x2962] session = 10dc :: [[SYN]] :: isn = 1c7e, options = 0021
[WARNING] OUTGOING:
SYN received in invalid state
[WARNING] INCOMING: [0x2962] session = 10dc :: [[SYN]] :: isn = 1c7e, options = 0021
[WARNING] OUTGOING:
SYN received in invalid state
[WARNING] INCOMING: [0x22a4] session = 10dc :: [[SYN]] :: isn = 1c7e, options = 0021
[WARNING] OUTGOING:
SYN received in invalid state
[WARNING] INCOMING: [0x22a4] session = 10dc :: [[SYN]] :: isn = 1c7e, options = 0021
[WARNING] OUTGOING:
dnscat [command: 4316]> SYN received in invalid state
[WARNING] INCOMING: [0x4cc5] session = 10dc :: [[SYN]] :: isn = 1c7e, options = 0021
[WARNING] OUTGOING:
SYN received in invalid state
[WARNING] INCOMING: [0x4cc5] session = 10dc :: [[SYN]] :: isn = 1c7e, options = 0021
[WARNING] OUTGOING:
SYN received in invalid state
[WARNING] INCOMING: [0x5bb1] session = 10dc :: [[SYN]] :: isn = 1c7e, options = 0021
[WARNING] OUTGOING:
SYN received in invalid state
[WARNING] INCOMING: [0x5bb1] session = 10dc :: [[SYN]] :: isn = 1c7e, options = 0021
[WARNING] OUTGOING:
```

Is there any other data you would want to analyse this? PCAPs etc?

If I set my DNS server to 8.8.8.8 connections are fine, however the current DNS resolution is:

VM -> VMWare Host -> Router -> ISP -> THE CLOUD

I'm sure something is probably mangling the request so its not in the format expected.

**Comment by iagox86** [comment]
That message happens often because of poor error handling. If the server
has an exception during the SYN processing, then it winds up in the wrong
state. I'm working hard on fixing it.

Can you try again with a version from today, if possible? I fixed a bunch
of things over the weekend that could cause that.

Ron

**Comment by Meatballs1** [comment]
This was a clone from about 30 minutes ago, I saw you had been updating :)

**Comment by iagox86** [comment]
Yeah, if you're going through the Internets, that's certainly not
surprising. I didn't get it into a state I was 100% comfortable with
yesterday, I need to track down what's going wrong still. Hopefully this
week. :)

**Comment by iagox86** [comment]
Hey, can you still reproduce this? I finally found/fixed a really, really subtle bug that was causing this, and that I've been banging my head against for weeks! That was the last thing on my TODO list before release, so I'm hoping this is related. :)

**Comment by Meatballs1** [comment]
I will try and take a look if I get a chance :) 

Is the fix server side or do I need to recompile client?

**Comment by iagox86** [comment]
I've made changes to both since then, so I recommend compiling the client
again. Make sure you "make clean", I spent too long tracking down a crash
caused by a stale .o file recently!

if it still happens, then the commands you run on the client and server
would be fantastic! Hopefully I can reproduce :)

**Comment by Meatballs1** [comment]
Tried a rebuild with the latest master as of today, but still get the SYN errors.

Also the client now spams:

`[[ ERROR ]] :: DNS: RCODE_SERVER_FAILURE`

**Comment by iagox86** [comment]
Make sure the server is listening and is on the right address/domain.

**Comment by iagox86** [comment]
Hey, I'm getting ready to release, the latest git build should be feature complete and error handling should be clean. Can you still reproduce this behavior?
